src/test/fixtures/mocks/window-navigation.js

```diff
@@ -3,37 +3,38 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 // @flow
-//
 import { coerceMatchingShape } from '../../../utils/flow';
 
+// This symbol will be used in the mock for window.location so that the mock for
+// window.history can change the inner location directly.
+const internalLocationAssign = Symbol.for('internalLocationAssign');
+
 /**
  * jsdom's implementation of this API doesn't allow for assigning to the location, which
  * we need to be able to do for certain tests.
  */
-export function mockWindowLocation(location: string = 'http://localhost') {
+function mockWindowLocation(location: string = 'http://localhost') {
   // This is the internal state.
   let url = new URL(location);
-  function setLocation(newUrl: string | { toString: () => string }): void {
-    location = newUrl.toString();
-    url = new URL(location);
+
+  function internalSetLocation(
+    newUrl: string | { toString: () => string }
+  ): void {
+    url = new URL(newUrl.toString(), url);
   }
 
   const nativeLocation = Object.getOwnPropertyDescriptor(window, 'location');
-  function accessError() {
-    throw new Error(
-      'Setting properties for the window.location object is not supported with this mock.'
-    );
-  }
 
   // It seems node v8 doesn't let us change the value unless we delete it before.
   delete window.location;
 
   const property = {
-    get(): $Shape<Location> {
```

Comment:
I remove the Shape because I added `searchParams` that's supported by both browsers and node, but Flow doesn't know about it sadly.

Reply:
This worked, thanks for the suggestion!

Reply:
And... actually I was confused. `searchParams` does exist in URL but not in Location. So I'm removing it!